src/mono/mono/metadata/assembly-load-context.c

```diff
@@ -84,6 +84,14 @@ mono_alcs_init (void)
 	default_alc->gchandle = mono_gchandle_new_internal (NULL, FALSE);
 }
 
+void
+mono_alc_foreach (MonoALCFunc func, gpointer user_data)
+{
+	alcs_lock ();
+	g_slist_foreach (alcs, func, user_data);
+	alcs_unlock ();
```

Comment:
I don't like that this creates a lock ordering dependency between the jit memory manager locks and the alcs lock.  It's not particularly obvious that those locks are related, so if another piece of code takes locks in a different order, we could get a classic deadlock.

What if `mono_alc_foreach` made a copy of the `alcs` list while holding the lock, and then called `g_slist_foreach` on the copy?

(And also we would need to protect each ALC, if it is collectible, from being collected while we're iterating)